pocs/sensors/arduino_io.py

```diff
@@ -0,0 +1,209 @@
+                records while reading from the serial line due to dropped
+                chars, and thus corrupted JSON.
+            serial_config:
+                The portion of the config file descibing the serial settings.
+        """
+        self.board_name = board_name or port
+        self.port = port
+        self._serial_data = open_serial_device(
+            port, serial_config=serial_config, name=board_name)
+        self._output_queue = output_queue
+        self._logger = get_root_logger()
+        self._last_reading = None
+
+        self._serial_data_lock = threading.Lock()
+        self._last_reading_lock = threading.Lock()
+        self._thread = None
+        self._stop = threading.Event()
+
+    def last_reading(self):
+        """Returns the last reading.
+
+        This may be newer than the last reading in the queue if the queue was full.
+        """
+        with self._last_reading_lock:
+            return self._last_reading
+
+    def start_reading(self):
+        if self._thread and self._thread.is_alive():
+            self._logger.debug('Thread {!r} is already running with ident {!r}',
+                               self._thread.name, self._thread.ident)
+            return
+
+        def reader():
+            self._logger.info('Started reader thread {!r} with ident {!r}',
+                              threading.current_thread().name, threading.get_ident())
+            announce = False
+            while not self._stop.is_set():
+                reading = self._read1()
+                if self._stop.is_set():
+                    break
+                if not reading:
+                    # Consider adding an error counter.
+                    announce = True
+                    continue
+                if announce:
+                    self._logger.info('Succeeded in reading from {!r}; got:\n{!r}',
+                                      self.port, reading)
+                    announce = False
+                self._handle_reading(reading)
+            self._logger.info('Stopping reader thread {!r} with ident {!r}',
+                              threading.current_thread().name, threading.get_ident())
+
+        self._thread = threading.Thread(target=reader, name=self.board_name)
+        self._thread.daemon = True
+        self._thread.start()
+
+    def stop_reading(self):
+        if self._thread and self._thread.is_alive():
+            self._logger.info('Waiting for thread {!r} to stop.', self._thread.name)
+            self._stop.set()
+            self._thread.join(timeout=4.0)
+            if self._thread.is_alive():
+                self._logger.error('Thread {!r} is still running.', self._thread.name)
+            else:
+                self._thread = None
+
+    def write(self, text):
+        with self._serial_data_lock:
+            if not self._serial_data.is_connected:
+                self._serial_data.connect()
+            return self._serial_data.write(text)
```

Comment:
I'm guessing the lock context manager does the right thing here with a return statement.

Reply:
Yes, it does.